### ISSUE TITLE
Crash on selecting share price in investment window (see Image)

### ISSUE THREAD

**Issue by siowena** [moneymanagerex/moneymanagerex#5618]
MMEX version:
 - [X] 1.6.3
Operating System:
 - [X] Mac OSX

# Description of the bug
Open portfolio view
Click on an investment
Add a share price (in image in comment below, this was the 02-02-2023 0.8386 row)
Close window and return to investment window
Click on the same investment
Click the share price just entered and it crashes the app (select 02-02-2023 row).
Clicking on an older share price the row highlights as expected (select 31-01-2023).

# Reproduction
Is the bug reproducible? 
 - [x] Always 
NOTE: I just updated to 1.6.3 before entering these values, the previous values where entered using 1.6.1 or 1.6.2 depending on age.

# Additional information
Looking into the DB, all the investment history records I've added for 2023 have a NULL in the HISTID. These are the records that crash application if selected in the share price history window.  If I edit the DB to add a HISTID seq number then start MMX and click on the investment share price it works OK.

Share prices that have been added via the auto-update feature seems to work as expected and not crash system. Only share prices added manually cause this issue.

NOTE: the sqlite_sequence table has no rows.

**Comment by siowena** [comment]
<img width="831" alt="Screen Shot 2023-02-06 at 22 00 06" src="https://user-images.githubusercontent.com/44472390/217050150-35e67055-1c62-40be-8c36-726de870b250.png">

Clicking on the 02-02-2023 row causes MMX to crash before adding a number into the HISTID column

**Comment by n-stein** [comment]
I am unable to reproduce a crash on Windows or Linux. @whalley, can you reproduce this on Mac?

**Comment by whalley** [comment]
Nope. I can't reproduce this crash. All seems to work OK.

**Comment by n-stein** [comment]
@siowena Can you try the latest Mac beta here https://link.ipx.co.uk/mmex-mac-betas to see if you still experience this issue? You can use a copy of your DB.

**Comment by siowena** [comment]
I'll look into this again and use the beta as well.

**Comment by siowena** [comment]
@n-stein Sorry but when I try to run the beta it gives me a mac security error and it can't check the download source and refuses to run the beta version.  Is there anyway to quickly get it to run?  I seem to remember I've got to change some systems settings but I'll have to dig out the details unless you have them to hand.

**Comment by whalley** [comment]
@siowena Which beta build are you using? They are all signed so should not cause security errors. Can you post a snapshot of the error message please.

**Comment by siowena** [comment]
mmex-1.6.4-Beta.2-Darwin-2023-02-23.dmg is the one I downloaded.

<img width="266" alt="Screen Shot 2023-03-02 at 10 07 09" src="https://user-images.githubusercontent.com/44472390/222345144-cd4b6b79-6d80-4a4e-bb01-0383d01a114b.png">

**Comment by siowena** [comment]
@whalley @n-stein 
OK, using the current versions 1.6.3 I created a new DB, new Investment account, new Stock and then update the prices many time manually.  Also did a autoupdate/download of prices. I can select the share price all cases without a crash.

It seems to be something caused by the upgrade of the DB and not a problem with the 1.6.3 code itself.

I can't tell the difference on the DB from 1.6.2 to 1.6.3 but I do see a sqlite_sequence table in my DB (the one upgraded) but don't see it in the new db created in 1.6.3

NOTE: It seems the HISTID is now missing from share prices downloaded automatically.
<img width="455" alt="Screen Shot 2023-03-02 at 13 16 47" src="https://user-images.githubusercontent.com/44472390/222385272-0d09ac4f-783a-406b-aa8f-591b7829975e.png">

**Comment by siowena** [comment]
@whalley  @n-stein 

I've noticed something else over time, the dates are repeating depending on the number of times I trigger the share price updates. It's as if it can't find the current date value (missing the HISTID value and so adds a new one).

How is the HIST ID calculated - is it a sequence (like oracle/ms-sql) or some other method?  As above, I don't see this on new DB for the latest build, just on the upgraded DB.

<img width="676" alt="Screen Shot 2023-03-14 at 14 51 16" src="https://user-images.githubusercontent.com/44472390/224978887-2f4388bb-31d9-478e-95f9-3162a4de171e.png">

**Comment by n-stein** [comment]
> I can't tell the difference on the DB from 1.6.2 to 1.6.3

There isn't one since there was no database update between 1.6.2 and 1.6.3. You can use them interchangeably with the same DB. If you open this DB on 1.6.2 do you see the same issue?

> How is the HIST ID calculated - is it a sequence (like oracle/ms-sql) or some other method?

HISTID is auto-assigned by the SQLite database engine, we don't specify the HISTID for inserts.

What is the output of `pragma table_info([stockhistory_v1]);`

**Comment by siowena** [comment]
@n-stein  output below
<img width="291" alt="Screen Shot 2023-03-16 at 14 20 30" src="https://user-images.githubusercontent.com/44472390/225587591-2a6e8d35-0cf3-42d3-a6d1-cc41a89a3d44.png">

**Comment by n-stein** [comment]
> Is it just a case of making a pk?

Yes, that's your issue. SQLite can't auto-generate the key because that table doesn't have a key field set. 

Unfortunately you can't just add a primary key, you'll have to setup a new table with the key then clone the data over. You don't need to delete the NULL records, SQLite will replace them in the new table with valid keys.

This code should do it:

```
CREATE TABLE "STOCKHISTORY_V1_TEMP" (
	"HISTID" integer primary key,
	"SYMBOL" TEXT NOT NULL,
	"DATE" TEXT NOT NULL,
	"VALUE" numeric NOT NULL,
	"UPDTYPE" integer,
	UNIQUE("SYMBOL","DATE")
);

INSERT INTO STOCKHISTORY_V1_TEMP SELECT * FROM STOCKHISTORY_V1;

DROP TABLE STOCKHISTORY_V1;

ALTER TABLE STOCKHISTORY_V1_TEMP RENAME TO STOCKHISTORY_V1;
```

**Comment by siowena** [comment]
@n-stein 
Thanks. I had to remove the 937 duplicate rows 1st, but now I've a system that doesn't crash when I select a stock price. I've updated a price manually and also triggered the auto-update. In both cases the prices appeared after the stop/start MMX and doesn't crash the program.

Looking the table, I can't find any HISTID = NULL rows, so seems all is solved.

I've a backup, so if I need a price from one of the duplicates I can look it up. Still not sure why the table would lose it's PK. Anyway, all good for now.

Thanks for the help, much appreciated.
